File: ios/ExpoNativeWechatModule.swift
import ExpoModulesCore

public class ExpoNativeWechatModule: Module {
    private var appid: String = ""
    
    private var logger = ExpoNativeWechatLogger(prefix: "[Native Wechat]")
    
    private func registerResponder () {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleResponse),
            name: Notification.Name("ResponseData"),
            object: nil
        )
    }
    
    @objc func handleResponse(_ notification: Notification) {
        let data = notification.userInfo as! [String : Any]
        
        sendEvent("ResponseFromNotification", data)
    }
    
    public func definition() -> ModuleDefinition {
        Name("ExpoNativeWechat")
        
        OnCreate {
            registerResponder()
        }
        
        Events("ResponseData")
        Events("ResponseFromNotification")
        
        Function("getConstants") {
            return [
                "WXSceneSession": WXSceneSession.rawValue,
                "WXSceneTimeline": WXSceneTimeline.rawValue,
                "WXSceneFavorite": WXSceneFavorite.rawValue,
                "WXMiniProgramTypeRelease": WXMiniProgramType.release.rawValue,
                "WXMiniProgramTypeTest": WXMiniProgramType.test.rawValue,
                "WXMiniProgramTypePreview": WXMiniProgramType.preview.rawValue,
            ]
        }
        
        Function("registerApp") {(params: RegisterAppParams) -> Void in
            if(params.log){
                if let logPrefix = params.logPrefix {
                    logger.logPrefix = logPrefix
                }
                
                WXApi.startLog(by: WXLogLevel.detail, logDelegate: logger)
            }
            
            appid = params.appid
            
            let success = WXApi.registerApp(appid, universalLink: params.universalLink)
            
            sendEvent("ResponseData", ["success": success, "id": params.id])
        }
        
        Function("checkUniversalLinkReady") { (params: BasicRequestParams) -> Void in
            WXApi.checkUniversalLinkReady { step, result in
                if result.success {
                    if step == WXULCheckStep.final {
                        self.sendEvent("ResponseData", [
                            "id": params.id,
                            "success": true,
                            "suggestion": "",
                            "errorInfo": ""
                        ])
                    }
                } else {
                    self.sendEvent("ResponseData", [
                        "id": params.id,
                        "success": true,
                        "suggestion": result.suggestion,
                        "errorInfo": result.errorInfo
                    ])
                }
            }
        }
        
        Function("isWechatInstalled") { (params: BasicRequestParams) -> Void in
            let installed = WXApi.isWXAppInstalled()
            
            sendEvent("ResponseData", ["success": installed, "id": params.id])
        }
        
        Function("sendAuthRequest") { (params: SendAuthRequestParams) -> Void in
            let req = SendAuthReq()
            
            if let scope = params.scope {
                req.scope = scope
            }
            
            req.state = params.state
            
            WXApi.send(req) { success in
                self.sendEvent("ResponseData", ["success": success, "id": params.id])
            }
        }
        
        Function("shareText") { (params: ShareTextParams) -> Void in
            let req = SendMessageToWXReq()
            
            req.bText = true
            req.text = params.text
            req.scene = Int32(params.scene)
            
            WXApi.send(req) { success in
                self.sendEvent("ResponseData", ["success": success, "id": params.id])
            }
        }
        
        Function("shareImage") { (params: ShareImageParams) -> Void in
            let url = URL(string: params.src);
            
            ExpoWechatUtils.downloadFile(url: url!) { data in
                guard data != nil else {
                    self.sendEvent("ResponseData", ["success": false, "id": params.id, "message": "Image data is empty"])
                    return
                }
                
                let imageObject = WXImageObject()
                imageObject.imageData = data!
                
                let message = WXMediaMessage()
                message.thumbData = data
                message.mediaObject = imageObject
                
                let req = SendMessageToWXReq()
                req.bText = false
                req.scene = Int32(params.scene)
                req.message = message
                
                WXApi.send(req) { success in
                    self.sendEvent("ResponseData", ["success": success, "id": params.id])
                }
            } onError: { error in
                self.sendEvent("ResponseData", ["success": false, "message": error?.localizedDescription])
            }
        }
        
        Function("shareVideo") { (params: ShareVideoParams) -> Void in
            let videoObj = WXVideoObject()
            
            videoObj.videoUrl = params.videoUrl
            videoObj.videoLowBandUrl = params.videoLowBandUrl ?? ""
            
            let message = WXMediaMessage()
            
            message.title = params.title ?? ""
            message.description = params.description ?? ""
            message.mediaObject = videoObj
            
            let onCoverDownloaded: ((Data?) -> Void)  = { data in
                if let imgData = data {
                    message.setThumbImage(UIImage(data: imgData)!)
                }
                
                let req = SendMessageToWXReq()
                req.bText = false
                req.message = message
                req.scene = Int32(params.scene)
                
                WXApi.send(req) { success in
                    self.sendEvent("ResponseData", ["success": success, "id": params.id])
                }
            }
            
            if params.coverUrl != nil {
                let url = URL(string: params.coverUrl!)
                
                ExpoWechatUtils.downloadFile(url: url!) { data in
                    if data == nil {
                        let compressed = ExpoWechatUtils.compressImage(data: data!, limit: 32000)
                        
                        onCoverDownloaded(compressed)
                    } else{
                        onCoverDownloaded(nil)
                    }
                } onError: { error in
                    self.sendEvent("ResponseData", ["id": params.id, "success": false, "message": error?.localizedDescription])
                }
                
            } else {
                onCoverDownloaded(nil)
            }
        }
        
        Function("shareWebpage") { (params: ShareWebpageParams) -> Void in
            let webpackObj = WXWebpageObject()
            
            webpackObj.webpageUrl = params.webpageUrl
            
            let message = WXMediaMessage()
            
            message.title = params.title ?? ""
            message.description = params.description ?? ""
            message.mediaObject = webpackObj
            
            let onCoverDownloaded: ((Data?) -> Void)  = { data in
                if let imgData = data {
                    message.setThumbImage(UIImage(data: imgData)!)
                }
                
                let req = SendMessageToWXReq()
                req.bText = false
                req.message = message
                req.scene = Int32(params.scene)
                
                WXApi.send(req) { success in
                    self.sendEvent("ResponseData", ["id": params.id, "success": success])
                }
            }
            
            if params.coverUrl != nil {
                let url = URL(string: params.coverUrl!)
                
                ExpoWechatUtils.downloadFile(url: url!) { data in
                    if data == nil {
                        let compressed = ExpoWechatUtils.compressImage(data: data!, limit: 32000)
                        
                        onCoverDownloaded(compressed)
                    } else{
                        onCoverDownloaded(nil)
                    }
                } onError: { error in
                    self.sendEvent("ResponseData", ["id": params.id, "success": false, "message": error?.localizedDescription])
                }
                
            } else {
                onCoverDownloaded(nil)
            }
        }
        
        Function("shareMiniProgram") { (params: ShareMiniProgramParams) -> Void in
            let object = WXMiniProgramObject()
            
            object.userName = params.userName
            object.webpageUrl = params.webpageUrl
            object.path = params.path
            object.withShareTicket = params.withShareTicket ?? false
            object.miniProgramType = WXMiniProgramType(rawValue: UInt(params.miniProgramType))!
            
            let message = WXMediaMessage()
            
            message.title = params.title ?? ""
            message.description = params.description ?? ""
            message.mediaObject = object
            
            let onCoverDownloaded: ((Data?) -> Void)  = { data in
                if let imgData = data {
                    message.setThumbImage(UIImage(data: imgData)!)
                }
                
                let req = SendMessageToWXReq()
                req.bText = false
                req.message = message
                req.scene = 0 // WXSceneSession
                
                WXApi.send(req) { success in
                    self.sendEvent("ResponseData", ["id": params.id, "success": success])
                }
            }
            
            if params.coverUrl != nil {
                let url = URL(string: params.coverUrl!)
                
                ExpoWechatUtils.downloadFile(url: url!) { data in
                    if data == nil {
                        let compressed = ExpoWechatUtils.compressImage(data: data!, limit: 32000)
                        
                        onCoverDownloaded(compressed)
                    } else{
                        onCoverDownloaded(nil)
                    }
                } onError: { error in
                    self.sendEvent("ResponseData", ["id": params.id, "success": false, "message": error?.localizedDescription])
                }
                
            } else {
                onCoverDownloaded(nil)
            }
        }
        
        Function("requestPayment") { (params: RequestPaymentParams) -> Void in
            let request = PayReq()
            
            request.partnerId = params.partnerId
            request.prepayId = params.prepayId
            request.package = "Sign=WXPay"
            request.nonceStr = params.nonceStr
            request.timeStamp = UInt32(params.timeStamp) ?? 0
            request.sign = params.sign
            
            
            WXApi.send(request) { success in
                self.sendEvent("ResponseData", ["id": params.id, "success": success])
            }
        }
        
        Function("requestSubscribeMessage") { (params: RequestSubscribeMsgParams) -> Void in
            var req = WXSubscribeMsgReq()
            
            req.scene = UInt32(params.scene)
            req.templateId = params.templateId
            req.reserved = params.reserved
            
            WXApi.send(req) { success in
                self.sendEvent("ResponseData", ["id": params.id, "success": success])
            }
        }
        
        Function("launchMiniProgram") { (params: LaunchMiniprogramParams) -> Void in
            var req = WXLaunchMiniProgramReq()
            
            req.userName = params.userName
            req.path = params.path
            req.miniProgramType = WXMiniProgramType(rawValue: UInt(UInt32(params.miniProgramType)))!
            
            WXApi.send(req) { success in
                self.sendEvent("ResponseData", ["id": params.id, "success": success])
            }
        }

        Function("openCustomerService") { (params: OpenCustomerServiceParams) -> Void in
            var req = WXOpenCustomerServiceReq()
            
            req.corpid = params.corpid
            req.url = params.url
            
            WXApi.send(req) { success in
                self.sendEvent("ResponseData", ["id": params.id, "success": success])
            }
        }
    }
}

File: android/src/main/java/expo/modules/nativewechat/ExpoNativeWechatModule.kt
package expo.modules.nativewechat

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Bitmap
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.core.os.bundleOf
import com.tencent.mm.opensdk.modelbase.BaseReq
import com.tencent.mm.opensdk.modelbase.BaseResp
import com.tencent.mm.opensdk.modelbiz.SubscribeMessage
import com.tencent.mm.opensdk.modelbiz.WXLaunchMiniProgram
import com.tencent.mm.opensdk.modelbiz.WXOpenCustomerServiceChat
import com.tencent.mm.opensdk.modelmsg.SendAuth
import com.tencent.mm.opensdk.modelmsg.SendMessageToWX
import com.tencent.mm.opensdk.modelmsg.WXImageObject
import com.tencent.mm.opensdk.modelmsg.WXMediaMessage
import com.tencent.mm.opensdk.modelmsg.WXMiniProgramObject
import com.tencent.mm.opensdk.modelmsg.WXTextObject
import com.tencent.mm.opensdk.modelmsg.WXVideoObject
import com.tencent.mm.opensdk.modelmsg.WXWebpageObject
import com.tencent.mm.opensdk.modelpay.PayReq
import com.tencent.mm.opensdk.openapi.IWXAPI
import com.tencent.mm.opensdk.openapi.IWXAPIEventHandler
import com.tencent.mm.opensdk.openapi.WXAPIFactory
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.nativewechat.ExpoNativeWechatUtils.DownloadBitmapCallback
import okhttp3.Call
import java.io.IOException


class ExpoNativeWechatModule : Module(), IWXAPIEventHandler {
  private var appid: String? = null
  private var registered = false
  private var wxApi: IWXAPI? = null
  private val REDIRECT_INTENT_ACTION = "com.hector.nativewechat.ACTION_REDIRECT_INTENT"
  private val reactContext
    get() = requireNotNull(appContext.reactContext)

  @RequiresApi(Build.VERSION_CODES.O)
  override fun definition() = ModuleDefinition {
    Name("ExpoNativeWechat")

    Function("getConstants") {
      return@Function bundleOf(
        "WXSceneSession" to SendMessageToWX.Req.WXSceneSession,
        "WXSceneTimeline" to SendMessageToWX.Req.WXSceneTimeline,
        "WXSceneFavorite" to SendMessageToWX.Req.WXSceneFavorite,
        "WXMiniProgramTypeRelease" to WXLaunchMiniProgram.Req.MINIPTOGRAM_TYPE_RELEASE,
        "WXMiniProgramTypeTest" to WXLaunchMiniProgram.Req.MINIPROGRAM_TYPE_TEST,
        "WXMiniProgramTypePreview" to WXLaunchMiniProgram.Req.MINIPROGRAM_TYPE_PREVIEW
      )
    }

    Events("ResponseData", "ResponseFromNotification")

    Function("registerApp") { params: RegisterAppParams ->
      appid = params.appid
      registered = true

      wxApi = WXAPIFactory.createWXAPI(reactContext, appid, true);

      val success = wxApi?.registerApp(appid)

      appContext.reactContext?.registerReceiver(object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
          handleIntent(intent.extras!!["intent"] as Intent?)
        }
      }, IntentFilter(REDIRECT_INTENT_ACTION), Context.RECEIVER_EXPORTED);

      this@ExpoNativeWechatModule.sendEvent(
        "ResponseData", bundleOf(
          "success" to success,
          "id" to params.id
        )
      )
    }

    Function("isWechatInstalled") { params: BasicRequestParams ->
      val installed = wxApi?.isWXAppInstalled()

      this@ExpoNativeWechatModule.sendEvent(
        "ResponseData", bundleOf(
          "success" to installed,
          "id" to params.id
        )
      )
    }

    Function("sendAuthRequest") { params: SendAuthRequestParams ->
      var req = SendAuth.Req()

      req.scope = params.scope
      req.state = req.state

      sendEvent(
        "ResponseData", bundleOf(
          "id" to params.id,
          "success" to wxApi?.sendReq(req)
        )
      )
    }

    Function("shareText") { params: ShareTextParams ->
      val text = params.text
      val scene = params.scene

      val textObj = WXTextObject()
      textObj.text = text

      val msg = WXMediaMessage()
      msg.mediaObject = textObj
      msg.description = text

      val req = SendMessageToWX.Req()
      req.message = msg
      req.scene = scene

      sendEvent(
        "ResponseData", bundleOf(
          "id" to params.id,
          "success" to wxApi?.sendReq(req)
        )
      )
    }

    Function("shareWebpage") { params: ShareWebpageParams ->
      val webpageUrl = params.webpageUrl
      val title = params.title
      val description = params.description
      val coverUrl = params.coverUrl
      val scene = params.scene

      val webpageObj = WXWebpageObject()
      webpageObj.webpageUrl = webpageUrl

      val msg = WXMediaMessage(webpageObj)
      msg.title = title
      msg.description = description

      val onCoverDownloaded: (Bitmap?) -> Unit = { bitmap ->
        bitmap?.let {
          msg.thumbData = ExpoNativeWechatUtils.bmpToByteArray(bitmap, 32, true)
        }

        val req = SendMessageToWX.Req()
        req.message = msg
        req.scene = scene

        sendEvent(
          "ResponseData", bundleOf(
            "id" to params.id,
            "success" to wxApi?.sendReq(req)
          )
        )
      }

      if (!coverUrl.isNullOrEmpty()) {
        ExpoNativeWechatUtils.downloadFileAsBitmap(coverUrl, object : DownloadBitmapCallback {
          override fun onFailure(call: Call, e: IOException) {
            this@ExpoNativeWechatModule.sendEvent(
              "ResponseData", bundleOf(
                "id" to params.id,
                "success" to false,
                "message" to e.localizedMessage
              )
            )
          }

          override fun onResponse(bitmap: Bitmap) {
            onCoverDownloaded(bitmap)
          }
        })
      } else {
        onCoverDownloaded(null)
      }
    }

    Function("shareVideo") { params: ShareVideoParams ->
      val videoUrl = params.videoUrl
      val videoLowBandUrl = params.videoLowBandUrl
      val title = params.title
      val description = params.description
      val coverUrl = params.coverUrl
      val scene = params.scene

      val video = WXVideoObject()
      video.videoUrl = videoUrl

      if (videoLowBandUrl != null) video.videoLowBandUrl = videoLowBandUrl

      val msg = WXMediaMessage(video)
      msg.title = title
      msg.description = description

      val onCoverDownloaded: (Bitmap?) -> Unit = { bitmap ->
        bitmap?.let {
          msg.thumbData = ExpoNativeWechatUtils.bmpToByteArray(bitmap, 32, true)
        }

        val req = SendMessageToWX.Req()
        req.message = msg
        req.scene = scene

        sendEvent(
          "ResponseData", bundleOf(
            "id" to params.id,
            "success" to wxApi?.sendReq(req)
          )
        )
      }

      if (!coverUrl.isNullOrEmpty()) {
        ExpoNativeWechatUtils.downloadFileAsBitmap(coverUrl, object : DownloadBitmapCallback {
          override fun onFailure(call: Call, e: IOException) {
            this@ExpoNativeWechatModule.sendEvent(
              "ResponseData", bundleOf(
                "id" to params.id,
                "success" to false,
                "message" to e.localizedMessage
              )
            )
          }

          override fun onResponse(bitmap: Bitmap) {
            onCoverDownloaded(bitmap)
          }
        })
      } else {
        onCoverDownloaded(null)
      }
    }

    Function("shareMiniProgram") { params: ShareMiniProgramParams ->
      val webpageUrl = params.webpageUrl
      val userName = params.userName
      val path = params.path
      val title = params.title
      val description = params.description
      val coverUrl = params.coverUrl
      val withShareTicket = params.withShareTicket
      val miniProgramType = params.miniProgramType
      val scene = params.scene

      val miniProgramObj = WXMiniProgramObject()
      miniProgramObj.webpageUrl = webpageUrl
      miniProgramObj.miniprogramType = miniProgramType
      miniProgramObj.userName = userName
      miniProgramObj.path = path
      miniProgramObj.withShareTicket = withShareTicket!!

      val msg = WXMediaMessage(miniProgramObj)
      msg.title = title
      msg.description = description

      val onCoverDownloaded: (Bitmap?) -> Unit = { bitmap ->
        bitmap?.let {
          msg.thumbData = ExpoNativeWechatUtils.bmpToByteArray(bitmap, 32, true)
        }

        val req = SendMessageToWX.Req()
        req.message = msg
        req.scene = scene

        sendEvent(
          "ResponseData", bundleOf(
            "id" to params.id,
            "success" to wxApi?.sendReq(req)
          )
        )
      }

      if (!coverUrl.isNullOrEmpty()) {
        ExpoNativeWechatUtils.downloadFileAsBitmap(coverUrl, object : DownloadBitmapCallback {
          override fun onFailure(call: Call, e: IOException) {
            this@ExpoNativeWechatModule.sendEvent(
              "ResponseData", bundleOf(
                "id" to params.id,
                "success" to false,
                "message" to e.localizedMessage
              )
            )
          }

          override fun onResponse(bitmap: Bitmap) {
            onCoverDownloaded(bitmap)
          }
        })
      } else {
        onCoverDownloaded(null)
      }
    }

    Function("requestPayment") { params: RequestPaymentParams ->
      val payReq = PayReq()

      payReq.partnerId = params.partnerId
      payReq.prepayId = params.prepayId
      payReq.nonceStr = params.nonceStr
      payReq.timeStamp = params.timeStamp
      payReq.sign = params.sign
      payReq.packageValue = "Sign=WXPay"
      payReq.extData = params.extData
      payReq.appId = appid

      sendEvent(
        "ResponseData", bundleOf(
          "id" to params.id,
          "success" to wxApi?.sendReq(payReq)
        )
      )
    }

    Function("requestSubscribeMessage") { params: RequestSubscribeMsgParams ->
      val templateId = params.templateId
      val reserved = params.reserved
      val scene = params.scene

      val req = SubscribeMessage.Req()
      req.scene = scene;
      req.templateID = templateId;
      req.reserved = reserved;

      sendEvent(
        "ResponseData", bundleOf(
          "id" to params.id,
          "success" to wxApi?.sendReq(req)
        )
      )
    }

    Function("launchMiniProgram") { params: LaunchMiniprogramParams ->
      val userName = params.userName
      val path = params.path
      val miniProgramType = params.miniProgramType

      val req = WXLaunchMiniProgram.Req()
      req.userName = userName;
      req.path = path;
      req.miniprogramType = miniProgramType;

      sendEvent(
        "ResponseData", bundleOf(
          "id" to params.id,
          "success" to wxApi?.sendReq(req)
        )
      )
    }

    Function("openCustomerService") { params: OpenCustomerServiceParams ->
      val corpId = params.corpid
      val url = params.url

      val req = WXOpenCustomerServiceChat.Req()
      req.corpId = corpId
      req.url = url

      sendEvent(
        "ResponseData", bundleOf(
          "id" to params.id,
          "success" to wxApi?.sendReq(req)
        )
      )
    }

  }

  fun handleIntent(intent: Intent?) {
    wxApi!!.handleIntent(intent, this)
  }

  override fun onReq(req: BaseReq?) {}

  override fun onResp(baseResp: BaseResp?) {
    val convertedData = NativeWechatRespDataHelper.downcastResp(baseResp)

    sendEvent("ResponseFromNotification", convertedData)
  }

}
